qt/DetailsDialog.cc

```diff
@@ -1289,38 +1290,64 @@ void DetailsDialog::onTrackerSelectionChanged()
 void DetailsDialog::onAddTrackerClicked()
 {
     bool ok = false;
-    QString const url = QInputDialog::getText(this, tr("Add URL "), tr("Add tracker announce URL:"), QLineEdit::Normal,
-        QString(), &ok);
+    QString const urls = QInputDialog::getMultiLineText(this, tr("Add URL(s) "),
+        tr("Add tracker announce URL(s) (one per line):"), {}, &ok);
+    QStringList urlList = urls.split(QRegExp(QString::fromStdString("\n|\r\n|\r")), QString::SkipEmptyParts);
+
+    // Filter out invalid urls
+    urlList.erase(std::remove_if(urlList.begin(),
+        urlList.end(),
+        [](auto const& url) { return !QUrl(url).isValid(); }),
+        urlList.end());
+
+    // Remove duplicate urls
+    urlList.sort();
+    urlList.erase(std::unique(urlList.begin(), urlList.end()),
+        urlList.end());
 
     if (!ok)
     {
         // user pressed "cancel" -- noop
     }
-    else if (!QUrl(url).isValid())
+    else if (urlList.isEmpty())
     {
-        QMessageBox::warning(this, tr("Error"), tr("Invalid URL \"%1\"").arg(url));
+        QMessageBox::warning(this, tr("Error"), tr("No valid URL(s) \"%1\"").arg(urls));
     }
     else
     {
-        torrent_ids_t ids;
+        QHash<QString, torrent_ids_t> torrentIdsPerUrl;
 
-        for (int const id : myIds)
+        for (auto const& url : urlList)
         {
-            if (myTrackerModel->find(id, url) == -1)
+            torrent_ids_t ids;
+
+            for (int const id : myIds)
             {
-                ids.insert(id);
+                if (myTrackerModel->find(id, url) == -1)
+                {
+                    ids.insert(id);
+                }
+            }
+
+            if (!ids.empty())
+            {
+                torrentIdsPerUrl[url] = ids;
             }
         }
 
-        if (ids.empty()) // all the torrents already have this tracker
+        if (torrentIdsPerUrl.empty()) // all the torrents already have this tracker
         {
-            QMessageBox::warning(this, tr("Error"), tr("Tracker already exists."));
+            QMessageBox::warning(this, tr("Error"), tr("Tracker(s) already exist."));
         }
```

Comment:
Here, we only need to check and see if there's *any* change -- as soon as we find any single diff, we can stop looking & don't need to keep walking through every combo.

It might be more readable to use extract-method here to clarify the test, putting a helper func into an anonymous namespace with something like

```cpp
namespace
{
  
bool torrentsHaveTrackers(TrackerModel const& tm, torrent_ids_t const& ids, QStringList const& urls)
{
    for (auto const& id : ids)
    {
        for (auto const & url : urls)
        {
            if (tm.find(id, url) == -1)
            {
                return false;
            }
        }
    }

    return true;
}

}
```

Then the code in this block could read:

```cpp
        auto const urlstrs = QUrl::toStringList(urls.toList());

        if (urlstrs.isEmpty() || torrentsHaveTrackers(*myTrackerModel, myIds, urlstrs))
        {
            QMessageBox::warning(this, tr("Error"), tr("No new URLs found."));
        }
```